src/algebra/lie_algebra.lean

```diff
@@ -0,0 +1,185 @@
+/-
+Copyright (c) 2019 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import ring_theory.algebra data.matrix.basic
+
+/-!
+# Lie algebras
+
+This file defines Lie rings, and Lie algebras over a commutative ring. It shows how these arise from
+associative rings and algebras via the ring commutator. In particular it defines the Lie algebra
+of endomorphisms of a module as well as of the algebra of square matrices over a commutative ring.
+
+## Notations
+
+We introduce the notation ⁅x, y⁆ for the Lie bracket. Note that these are the Unicode "square with
+quill" square brackets rather than the usual square brackets.
+
+## Implementation notes
+
+Lie algebras are defined as modules with a compatible Lie ring structure, and thus are partially
+unbundled. Since they extend Lie rings, these are also partially unbundled.
+
+## References
+* [N. Bourbaki, *Lie Groups and Lie Algebras, Chapters 1--3*][bourbaki1975]
+
+## Tags
+
+lie bracket, ring commutator, jacobi identity, lie ring, lie algebra
+-/
+
+universes u v
+
+class has_bracket (L : Type v) := (bracket : L → L → L)
+
+notation `⁅`x`,` y`⁆` := has_bracket.bracket x y
+
+namespace ring_commutator
+
+variables (A : Type v) [ring A]
+
+instance : has_bracket A :=
```

Comment:
This instance might cause problems for classes of `add_comm_groups` that are sometimes rings. ` If I had a family of type `X i`, where `X i` was a ring for some `i`, and a `lie_algebra` for all `i`, then there would be two `has_bracket` instances on the `i` such that `X i` is a `ring`, and it would be important to make sure they were defeq.

`matrix` is an example of a family that is a `ring` for some argument, though I don't know if there are any examples that would cause a problem for this instance. @jcommelin what do you think of this?

Reply:
Hmmm, I see. I guess this could at some point give us headaches. On possibility is to use a type wrapper here. @rwbarton You usually seem to have a clear view on these things. What do you suggest?

Reply:
I have just pushed another commit which attempts to address this concern. Thanks for raising such an insightful issue with my previous approach.